File: packages/shared/src/schemas/application.ts
import { z } from 'zod';

export const ValidationSchema = z.object({
  pattern: z.string().optional(),
  min: z.number().optional(),
  max: z.number().optional(),
});

export const SelectOptionSchema = z.object({
  value: z.string(),
  label: z.string(),
});

export const ApplicationFieldTypeSchema = z.enum([
  'text',
  'textarea',
  'email',
  'tel',
  'url',
  'select',
  'radio',
  'checkbox',
  'date',
  'file',
  'unknown',
]);

export const ApplicationFieldSchema = z.object({
  key: z.string(),
  label: z.string(),
  type: ApplicationFieldTypeSchema,
  required: z.boolean(),
  options: z.array(SelectOptionSchema).optional(),
  validation: ValidationSchema.optional(),
  source_hint: z.string().optional(),
});

export type Validation = z.infer<typeof ValidationSchema>;
export type SelectOption = z.infer<typeof SelectOptionSchema>;
export type ApplicationFieldType = z.infer<typeof ApplicationFieldTypeSchema>;
export type ApplicationField = z.infer<typeof ApplicationFieldSchema>;

File: apps/api/app/schemas/application.py
from pydantic import BaseModel
from typing import Optional, List
from enum import Enum


class ApplicationFieldType(str, Enum):
    text = "text"
    textarea = "textarea"
    email = "email"
    tel = "tel"
    url = "url"
    select = "select"
    radio = "radio"
    checkbox = "checkbox"
    date = "date"
    file = "file"
    unknown = "unknown"


class Validation(BaseModel):
    pattern: Optional[str] = None
    min: Optional[int] = None
    max: Optional[int] = None


class SelectOption(BaseModel):
    value: str
    label: str


class ApplicationField(BaseModel):
    key: str
    label: str
    type: ApplicationFieldType
    required: bool
    options: Optional[List[SelectOption]] = None
    validation: Optional[Validation] = None
    source_hint: Optional[str] = None
